### DISCUSSION TITLE
Helm release failed due to " can't access 'GitRepository', cross-namespace references have been blocked"

### DISCUSSION THREAD

**Discussion by dongfrank** [fluxcd/flux2#4216] [Q&A] [answered]
Here is the gitrepo :
`
apiVersion: source.toolkit.fluxcd.io/v1
kind: GitRepository
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: >
      {"apiVersion":"source.toolkit.fluxcd.io/v1beta2","kind":"GitRepository","metadata":{"annotations":{},"name":"sre","namespace":"sre"},"spec":{"interval":"1m0s","ref":{"branch":"ARTMC-14620_remove-admin-rights-from-flux","tag":null},"secretRef":{"name":"sre-git-secret"},"timeout":"60s","url":"https://github.boschdevcloud.com/MC/cloudrangers-additional-environment-configurations.git"}}
  creationTimestamp: '2023-09-05T06:06:14Z'
  finalizers:
    - finalizers.fluxcd.io
  generation: 1
  managedFields:
    - apiVersion: source.toolkit.fluxcd.io/v1beta2
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:spec:
          .: {}
          f:gitImplementation: {}
          f:interval: {}
          f:ref:
            .: {}
            f:branch: {}
          f:secretRef:
            .: {}
            f:name: {}
          f:timeout: {}
          f:url: {}
      manager: HashiCorp
      operation: Update
      time: '2023-09-05T06:06:14Z'
    - apiVersion: source.toolkit.fluxcd.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:finalizers:
            .: {}
            v:"finalizers.fluxcd.io": {}
      manager: source-controller
      operation: Update
      time: '2023-09-05T06:06:14Z'
    - apiVersion: source.toolkit.fluxcd.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:artifact:
            .: {}
            f:digest: {}
            f:lastUpdateTime: {}
            f:path: {}
            f:revision: {}
            f:size: {}
            f:url: {}
          f:conditions: {}
          f:observedGeneration: {}
      manager: source-controller
      operation: Update
      subresource: status
      time: '2023-09-05T08:56:33Z'
  name: sre
  namespace: sre
  resourceVersion: '93269'
  uid: d5828d79-a733-4d6d-a2ed-0f8774311d9f
  selfLink: /apis/source.toolkit.fluxcd.io/v1/namespaces/sre/gitrepositories/sre
status:
  artifact:
    digest: sha256:504e038f99067dd8b58d9b4dfec9deeff2fb625477f30253fbc29cf495c7974d
    lastUpdateTime: '2023-09-05T08:56:33Z'
    path: gitrepository/sre/sre/fe79d6a55cb5f20d91ba59e98a570ae5aec25b26.tar.gz
    revision: >-
      ARTMC-14620_remove-admin-rights-from-flux@sha1:fe79d6a55cb5f20d91ba59e98a570ae5aec25b26
    size: 275271
    url: >-
      http://source-controller.flux-system.svc.cluster.local./gitrepository/sre/sre/fe79d6a55cb5f20d91ba59e98a570ae5aec25b26.tar.gz
  conditions:
    - lastTransitionTime: '2023-09-05T08:56:33Z'
      message: >-
        stored artifact for revision
        'ARTMC-14620_remove-admin-rights-from-flux@sha1:fe79d6a55cb5f20d91ba59e98a570ae5aec25b26'
      observedGeneration: 1
      reason: Succeeded
      status: 'True'
      type: Ready
    - lastTransitionTime: '2023-09-05T08:56:33Z'
      message: >-
        stored artifact for revision
        'ARTMC-14620_remove-admin-rights-from-flux@sha1:fe79d6a55cb5f20d91ba59e98a570ae5aec25b26'
      observedGeneration: 1
      reason: Succeeded
      status: 'True'
      type: ArtifactInStorage
  observedGeneration: 1
spec:
  interval: 1m0s
  ref:
    branch: ARTMC-14620_remove-admin-rights-from-flux
  secretRef:
    name: sre-git-secret
  timeout: 60s
  url: >-
    https://github.boschdevcloud.com/MC/cloudrangers-additional-environment-configurations.git

`

here is the helm release 
`
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  creationTimestamp: '2023-09-05T03:02:16Z'
  finalizers:
    - finalizers.fluxcd.io
  generation: 1
  labels:
    kustomize.toolkit.fluxcd.io/name: sre
    kustomize.toolkit.fluxcd.io/namespace: sre
  managedFields:
    - apiVersion: helm.toolkit.fluxcd.io/v2beta1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:labels:
            f:kustomize.toolkit.fluxcd.io/name: {}
            f:kustomize.toolkit.fluxcd.io/namespace: {}
        f:spec:
          f:chart:
            f:spec:
              f:chart: {}
              f:sourceRef:
                f:kind: {}
                f:name: {}
                f:namespace: {}
          f:dependsOn: {}
          f:install:
            f:remediation:
              f:retries: {}
          f:interval: {}
          f:releaseName: {}
          f:serviceAccountName: {}
          f:targetNamespace: {}
          f:values:
            f:certificate:
              .: {}
              f:name: {}
              f:namespace: {}
              f:secretName: {}
            f:createCert: {}
          f:valuesFrom: {}
      manager: kustomize-controller
      operation: Apply
      time: '2023-09-05T03:02:16Z'
    - apiVersion: helm.toolkit.fluxcd.io/v2beta1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:finalizers:
            .: {}
            v:"finalizers.fluxcd.io": {}
      manager: Go-http-client
      operation: Update
      time: '2023-09-05T03:02:16Z'
    - apiVersion: helm.toolkit.fluxcd.io/v2beta1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:conditions: {}
          f:failures: {}
          f:observedGeneration: {}
      manager: helm-controller
      operation: Update
      subresource: status
      time: '2023-09-05T03:02:16Z'
  name: virtual-env-cert
  namespace: blueprint
  resourceVersion: '968044'
  uid: 26994a36-633c-4485-99b9-56fc73837982
  selfLink: >-
    /apis/helm.toolkit.fluxcd.io/v2beta1/namespaces/blueprint/helmreleases/virtual-env-cert
status:
  conditions:
    - lastTransitionTime: '2023-09-05T03:02:16Z'
      message: >-
        can't access 'GitRepository/sre/sre', cross-namespace references have
        been blocked
      reason: AccessDenied
      status: 'False'
      type: Ready
  failures: 1
  observedGeneration: 1
spec:
  chart:
    spec:
      chart: ./charts/istio-utils
      reconcileStrategy: ChartVersion
      sourceRef:
        kind: GitRepository
        name: sre
        namespace: sre
      version: '*'
  dependsOn:
    - name: cert-manager
  install:
    remediation:
      retries: 5
  interval: 1m
  releaseName: virtual-env-cert
  serviceAccountName: flux-admin
  targetNamespace: istio-ingress
  values:
    certificate:
      name: ingress-cert
      namespace: istio-ingress
      secretName: ingress-tls
    createCert: true
  valuesFrom:
    - kind: ConfigMap
      name: istio-gateway-config
      optional: false
      valuesKey: baseUrl

`

here is the service account and cluster role bindings
`apiVersion: v1
kind: ServiceAccount
metadata:
  name: flux-admin
  namespace: blueprint
  uid: e2376c13-f892-4d9d-9f55-a934bb309c24
  resourceVersion: '3573'
  creationTimestamp: '2023-09-05T05:32:57Z'
  managedFields:
    - manager: HashiCorp
      operation: Update
      apiVersion: v1
      time: '2023-09-05T05:32:57Z'
      fieldsType: FieldsV1
      fieldsV1:
        f:automountServiceAccountToken: {}
  selfLink: /api/v1/namespaces/blueprint/serviceaccounts/flux-admin
automountServiceAccountToken: true

apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: flux-admin-binding
  uid: 377284c7-1b34-433b-a9d2-d3f106dc5355
  resourceVersion: '3575'
  creationTimestamp: '2023-09-05T05:32:58Z'
  managedFields:
    - manager: HashiCorp
      operation: Update
      apiVersion: rbac.authorization.k8s.io/v1
      time: '2023-09-05T05:32:58Z'
      fieldsType: FieldsV1
      fieldsV1:
        f:roleRef: {}
        f:subjects: {}
  selfLink: /apis/rbac.authorization.k8s.io/v1/clusterrolebindings/flux-admin-binding
subjects:
  - kind: ServiceAccount
    name: flux-admin
    namespace: blueprint
  - kind: ServiceAccount
    name: flux-admin
    namespace: flux-system
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
`
Even I granted a cluster admin role for service account which is used in helmrelease, it still failed due to the cross namespace  policy . I want to know why it happened, did I miss something and what to do

**Comment by somtochiama** [comment, marked as answer]
Hey, Disabling cross namespace reference is set by the controller flag `--no-cross-namespace-refs`.
This is done to deploy Flux in a multi-tenant manner and disable access across tenants.

Please see docs:
https://fluxcd.io/flux/security/#cross-namespace-reference-policy
https://fluxcd.io/flux/components/helm/helmreleases/#helm-chart-template

**Reply by dongfrank** [reply]
Thanks for your reply and I want to know even if I assign a cluster admin service account to helm release, it still can't access the gitrepo from other namespace ?

**Reply by somtochiama** [reply]
No, that will not resolve it. Removing the flag will. 
You might need to remove a patch from the kustomization.yaml in your bootstrap repo depending on how Flux was installed. Please see the docs above.

**Reply by dongfrank** [reply]
OK, I removed the `--no-cross-namespace-refs` and it works. thank you for your help!
